File: frontend/src/stores/teachingPlanWorkbench.ts
import { defineStore } from 'pinia'
import http, { getTeacherIdentity, teacherIdentityHeaders } from '../utils/http'
import { createUuid } from '../utils/client-id'
import type { CourseTeachingPlanProjection } from './types'
import { postGenerationStream } from '../shared/generation-stream'

export interface TeachingPlanOperation {
  operation_id: string
  path: string
  before: unknown
  after: unknown
  source?: 'manual' | 'ai' | 'restore'
}

const teacherRequestConfig = <T extends Record<string, unknown>>(extra?: T) => ({
  headers: { 'X-User-Id': getTeacherIdentity() },
  ...(extra || {}),
})

export interface TeachingPlanDraft {
  draft_id: string
  base_plan_revision_id: string
  base_course_document_revision: string
  changed_paths: string[]
  operations: TeachingPlanOperation[]
  validation?: TeachingPlanValidation
  updated_at?: string
}

export interface TeachingPlanValidation {
  passed: boolean
  issues?: Array<{ code?: string; message?: string; severity?: string }>
  blockers?: Array<{ code?: string; message?: string; severity?: string }>
}

export interface TeachingPlanImpactItem {
  group?: string
  type: string
  id: string
  reason: string
  status?: string
}

export interface TeachingPlanImpact {
  changed: TeachingPlanImpactItem[]
  needs_regeneration: TeachingPlanImpactItem[]
  stale: TeachingPlanImpactItem[]
  unchanged: TeachingPlanImpactItem[]
  blocked: TeachingPlanImpactItem[]
  blocking: boolean
}

export interface TeachingPlanChangeSet {
  change_set_id: string
  draft_id: string
  status: 'draft' | 'ready' | 'blocked' | 'stale' | 'superseded' | 'applied' | 'rejected'
  operations: TeachingPlanOperation[]
  validation?: TeachingPlanValidation
  impact_report?: TeachingPlanImpact
  created_at?: string
  applied_revision_id?: string
}

export interface TeachingPlanAICandidate {
  candidate_id: string
  draft_id: string
  status: 'ready' | 'accepted' | 'rejected' | 'stale'
  rationale?: string
  operations: TeachingPlanOperation[]
  validation?: TeachingPlanValidation
  impact_report?: TeachingPlanImpact
  created_at?: string
}

export interface TeachingPlanRevision {
  revision_id: string
  revision_number: number
  parent_revision_id?: string
  restored_from_revision_id?: string
  created_by?: string
  created_at?: string
  quality_report?: TeachingPlanValidation
}

export interface TeachingPlanWorkbench {
  course_id: string
  enabled: boolean
  available: boolean
  can_initialize: boolean
  read_only_reason: string
  current_plan_revision_id: string
  course_document_revision: string
  teaching_plan: CourseTeachingPlanProjection
  draft: TeachingPlanDraft | null
  revisions: TeachingPlanRevision[]
  change_sets: TeachingPlanChangeSet[]
  ai_candidates: TeachingPlanAICandidate[]
  // 后端 field_permission 返回三态；分小节的字段几乎都是
  // requires_impact_review（可写草稿，但应用前必须有完整影响报告）。
  editable_fields: Array<{
    path: string
    state: 'editable' | 'requires_impact_review' | 'readonly'
    reason: string
    value_hash?: string
  }>
  // 学科模板为每节提供的候选教学环节：前端据此渲染增删入口。
  // 必需环节由模板规定，取消勾选会被后端的必需环节合同拒绝。
  section_module_options?: Record<string, Array<{
    module_id: string
    label: string
    required: boolean
    selected: boolean
    output_contract: string
  }>>
  downstream: Record<string, unknown>
}

export interface TeachingPlanReview {
  draft_id: string
  base_plan_revision_id: string
  diff: { operations: TeachingPlanOperation[] }
  impact_report: TeachingPlanImpact
  validation: TeachingPlanValidation
}

const requestId = (prefix: string) => (
  `${prefix}_${createUuid()}`
)

function apiErrorCode(error: any): string {
  if (typeof error?.code === 'string') return error.code
  const detail = error?.response?.data?.detail
  if (detail && typeof detail === 'object' && typeof detail.code === 'string') return detail.code
  return error?.response?.status === 409 ? 'teaching_plan_conflict' : 'teaching_plan_request_failed'
}

function apiErrorMessage(error: any): string {
  if (typeof error?.message === 'string' && error?.detail) return error.message
  const detail = error?.response?.data?.detail
  return detail && typeof detail === 'object' && typeof detail.message === 'string'
    ? detail.message
    : ''
}

// 领域错误的 details 里带着可执行信息（例如 redirect_to_outline_edit 的目录
// 编辑器 endpoint）。只记 code 会把这些信息丢掉，前端就只能显示一句文案、
// 没法真的把教师送过去。
function apiErrorDetail(error: any): Record<string, unknown> {
  if (error?.detail && typeof error.detail === 'object') return { ...error.detail }
  const detail = error?.response?.data?.detail
  return detail && typeof detail === 'object' ? { ...detail } : {}
}

export const useTeachingPlanWorkbenchStore = defineStore('teachingPlanWorkbench', {
  state: () => ({
    courseId: '',
    workbench: null as TeachingPlanWorkbench | null,
    review: null as TeachingPlanReview | null,
    revisionDiff: null as {
      left_revision_id: string
      right_revision_id: string
      diff: { operations: TeachingPlanOperation[] }
    } | null,
    loading: false,
    savingPaths: [] as string[],
    pendingAction: '' as '' | 'confirmSource' | 'initialize' | 'ai' | 'review' | 'apply' | 'restore' | 'discard',
    errorCode: '',
    errorMessage: '',
    errorDetail: {} as Record<string, unknown>,
    generationMessage: '',
  }),

  getters: {
    draft: state => state.workbench?.draft || null,
    isSaving: state => state.savingPaths.length > 0,
  },

  actions: {
    reset() {
      this.courseId = ''
      this.workbench = null
      this.review = null
      this.revisionDiff = null
      this.savingPaths = []
      this.pendingAction = ''
      this.generationMessage = ''
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
    },

    applyWorkbench(workbench: TeachingPlanWorkbench) {
      if (this.courseId && this.courseId !== workbench.course_id) {
        this.review = null
        this.revisionDiff = null
        this.savingPaths = []
      }
      this.workbench = workbench
      this.courseId = workbench.course_id
    },

    async load(courseId: string) {
      if (!courseId) return this.reset()
      this.loading = true
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.get(`/api/courses/${courseId}/teaching-plan/workbench`, teacherRequestConfig({ silentError: true }))
        this.applyWorkbench(data.workbench)
        return this.workbench
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.loading = false
      }
    },

    async confirmGenerationPreview(courseId: string, sourceTaskId: string) {
      if (!courseId) return null
      this.pendingAction = 'confirmSource'
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/teacher/courses/${courseId}/authoring/confirm-generation-preview`,
          { confirm: true, source_task_id: sourceTaskId },
          teacherRequestConfig({ silentError: true }),
        )
        await this.load(courseId)
        return data
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async initializeBaseline() {
      if (!this.workbench || !this.courseId) return null
      this.pendingAction = 'initialize'
      this.errorCode = ''
      this.errorMessage = ''
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/baseline`,
          {
            base_course_document_revision: this.workbench.course_document_revision,
            idempotency_key: requestId('initialize_plan'),
          },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        return data.receipt || null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async beginDraft() {
      if (!this.workbench || !this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/drafts`,
          {
            base_plan_revision_id: this.workbench.current_plan_revision_id,
            base_course_document_revision: this.workbench.course_document_revision,
            idempotency_key: requestId('create_draft'),
          },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        this.review = null
        return this.draft
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      }
    },

    async patchDraft(path: string, value: unknown, expectedValueHash?: string) {
      const draft = this.draft
      if (!draft || !this.workbench || !this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      this.savingPaths = [...new Set([...this.savingPaths, path])]
      try {
        const { data } = await http.patch(
          `/api/courses/${this.courseId}/teaching-plan/drafts/${draft.draft_id}`,
          {
            path,
            value,
            expected_value_hash: expectedValueHash ?? (
              this.workbench.editable_fields.find(field => field.path === path)?.value_hash || ''
            ),
            base_plan_revision_id: draft.base_plan_revision_id,
            idempotency_key: requestId('patch_draft'),
          },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        this.review = null
        return this.draft
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.savingPaths = this.savingPaths.filter(item => item !== path)
      }
    },

    async reviewDraft() {
      const draft = this.draft
      if (!draft || !this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'review'
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/validate`,
          { draft_id: draft.draft_id, idempotency_key: requestId('validate_draft') },
          teacherRequestConfig({ silentError: true }),
        )
        this.review = data.review
        return this.review
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async discardDraft() {
      const draft = this.draft
      if (!draft || !this.courseId) return
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'discard'
      this.errorDetail = {}
      try {
        const { data } = await http.delete(
          `/api/courses/${this.courseId}/teaching-plan/drafts/${draft.draft_id}`,
          teacherRequestConfig({
            data: { idempotency_key: requestId('discard_draft') },
            silentError: true,
          }),
        )
        this.applyWorkbench(data.workbench)
        this.review = null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async createAiCandidate(paths: string[], instruction: string) {
      const draft = this.draft
      if (!draft || !this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'ai'
      this.errorDetail = {}
      this.generationMessage = ''
      try {
        const data = await postGenerationStream<{ workbench: TeachingPlanWorkbench }>(
          `/api/courses/${this.courseId}/teaching-plan/drafts/${draft.draft_id}/ai-candidates`,
          { paths, instruction, idempotency_key: requestId('ai_candidate') },
          {
            headers: teacherIdentityHeaders(),
            onProgress: progress => {
              this.generationMessage = progress.message || ''
            },
          },
        )
        this.applyWorkbench(data.workbench)
        return this.workbench?.ai_candidates.find(item => item.status === 'ready') || null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
        this.generationMessage = ''
      }
    },

    async acceptAiCandidate(candidateId: string, operationIds: string[]) {
      if (!this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/ai-candidates/${candidateId}/accept`,
          { operation_ids: operationIds, idempotency_key: requestId('ai_accept') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        return this.draft
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      }
    },

    async rejectAiCandidate(candidateId: string) {
      if (!this.courseId) return
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/ai-candidates/${candidateId}/reject`,
          { idempotency_key: requestId('ai_reject') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      }
    },

    async loadRevisionDiff(revisionId: string) {
      if (!this.courseId || !this.workbench) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.get(
          `/api/courses/${this.courseId}/teaching-plan/revisions/${revisionId}/diff`,
          teacherRequestConfig({
            params: { against: this.workbench.current_plan_revision_id },
            silentError: true,
          }),
        )
        this.revisionDiff = data
        return this.revisionDiff
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      }
    },

    async restoreRevision(revisionId: string) {
      if (!this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'restore'
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/revisions/${revisionId}/restore`,
          { idempotency_key: requestId('restore_revision') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        this.revisionDiff = null
        return data.receipt || null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async createChangeSet() {
      const draft = this.draft
      if (!draft || !this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'review'
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/change-sets`,
          { draft_id: draft.draft_id, idempotency_key: requestId('create_change_set') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        const changeSet = this.workbench?.change_sets.find(item => item.draft_id === draft.draft_id)
        if (changeSet) {
          this.review = {
            draft_id: draft.draft_id,
            base_plan_revision_id: draft.base_plan_revision_id,
            diff: { operations: changeSet.operations || [] },
            impact_report: changeSet.impact_report || emptyImpact(),
            validation: changeSet.validation || { passed: false },
          }
        }
        return changeSet || null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async applyChangeSet(changeSetId: string) {
      if (!this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.pendingAction = 'apply'
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/change-sets/${changeSetId}/apply`,
          { idempotency_key: requestId('apply_change_set') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        this.review = null
        return data.receipt || null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      } finally {
        this.pendingAction = ''
      }
    },

    async rejectChangeSet(changeSetId: string) {
      if (!this.courseId) return null
      this.errorCode = ''
      this.errorMessage = ''
      this.errorDetail = {}
      try {
        const { data } = await http.post(
          `/api/courses/${this.courseId}/teaching-plan/change-sets/${changeSetId}/reject`,
          { idempotency_key: requestId('reject_change_set') },
          teacherRequestConfig({ silentError: true }),
        )
        this.applyWorkbench(data.workbench)
        this.review = null
      } catch (error) {
        this.errorCode = apiErrorCode(error)
        this.errorMessage = apiErrorMessage(error)
        this.errorDetail = apiErrorDetail(error)
        throw error
      }
    },
  },
})

function emptyImpact(): TeachingPlanImpact {
  return {
    changed: [],
    needs_regeneration: [],
    stale: [],
    unchanged: [],
    blocked: [],
    blocking: false,
  }
}

File: backend/routers/teaching_plan_workbench.py
"""HTTP boundary for structured teaching-plan drafts and revisions."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel, Field

from course_repository import CourseDocumentConflict, CourseDocumentNotFound
from dependencies import get_course_document_repository
from learner_context import resolve_user_id
from teaching_plan_workbench import (
    TeachingPlanWorkbenchError,
    TeachingPlanWorkbenchService,
)
from teaching_representations import teaching_representation_repository


router = APIRouter(
    prefix="/courses/{course_id}/teaching-plan",
    tags=["teaching_plan_workbench"],
)


class DraftRequest(BaseModel):
    base_plan_revision_id: str = Field(default="", max_length=200)
    base_course_document_revision: str = Field(default="", max_length=200)
    idempotency_key: str = Field(min_length=1, max_length=200)


class PatchDraftRequest(BaseModel):
    path: str = Field(min_length=1, max_length=500)
    value: Any
    expected_value_hash: str = Field(default="", max_length=200)
    base_plan_revision_id: str = Field(default="", max_length=200)
    idempotency_key: str = Field(min_length=1, max_length=200)


class ChangeSetRequest(BaseModel):
    draft_id: str = Field(min_length=1, max_length=200)
    idempotency_key: str = Field(min_length=1, max_length=200)


class CommandRequest(BaseModel):
    idempotency_key: str = Field(min_length=1, max_length=200)


class InitializeBaselineRequest(CommandRequest):
    base_course_document_revision: str = Field(default="", max_length=200)


class AICandidateRequest(BaseModel):
    # 上限按「一个小节的全部可编辑字段」定，不是按人手勾选的字段数定：
    # 需求 5 的分小节优化会把整节路径一次性发出，而一节的路径数随环节数与
    # 知识点数增长（10 条小节字段 + 1 条环节顺序 + 环节×5 + 知识点×7）。
    # 真实课程实测：矩阵与线性变换 12 个小节，每节 75–89 条，最多的一节
    # 5 个环节 7 个知识点共 84 条。合成夹具只有 17 条，据此定的上限会在
    # 真实数据上继续 422，所以这里按真实规模留足余量。
    # 领域层仍逐条校验白名单与只读，上限只防御性地挡住畸形请求。
    paths: list[str] = Field(min_length=1, max_length=256)
    instruction: str = Field(min_length=1, max_length=3000)
    idempotency_key: str = Field(min_length=1, max_length=200)


class CandidateCommandRequest(CommandRequest):
    # 逐项接受 AI 候选时同样可能覆盖整节，与 paths 对齐。
    operation_ids: list[str] = Field(default_factory=list, max_length=256)


def _service(repository=Depends(get_course_document_repository)) -> TeachingPlanWorkbenchService:
    # 表达注册表只读接入：影响分析据此把 needs_regeneration 收窄到真实引用的
    # 下游对象；拿不到注册表时分析自动退回按小节的保守答案，不阻断工作台。
    return TeachingPlanWorkbenchService(
        repository,
        representation_repository=teaching_representation_repository,
    )


def _actor(request: Request) -> str:
    return resolve_user_id(request.headers.get("X-User-Id"))


def _error(exc: Exception) -> HTTPException:
    if isinstance(exc, TeachingPlanWorkbenchError):
        status_code = 409 if exc.code.endswith("conflict") or exc.code.endswith("blocked") else 400
        if exc.code.endswith("not_found") or exc.code == "teaching_plan_missing":
            status_code = 404
        # 结构操作重定向不是「请求错误」，是「这件事归目录真源管」：
        # 前端据此跳目录编辑器，语义与版本冲突同属 409。
        if exc.code == "redirect_to_outline_edit":
            status_code = 409
        return HTTPException(status_code=status_code, detail={
            "code": exc.code,
            "message": str(exc),
            **exc.details,
        })
    if isinstance(exc, CourseDocumentNotFound):
        return HTTPException(status_code=404, detail={"code": "course_not_found", "message": "课程不存在。"})
    if isinstance(exc, CourseDocumentConflict):
        return HTTPException(status_code=409, detail={"code": "course_document_conflict", "message": str(exc)})
    return HTTPException(status_code=500, detail={"code": "teaching_plan_workbench_error", "message": str(exc)})


@router.get("/workbench")
async def get_workbench(
    course_id: str,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {"status": "success", "workbench": service.view(course_id, actor=_actor(request))}
    except Exception as exc:  # Convert domain conflicts into a stable API contract.
        raise _error(exc) from exc


@router.post("/baseline")
async def initialize_baseline(
    course_id: str,
    body: InitializeBaselineRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "initialized",
            **await service.initialize_baseline(
                course_id,
                actor=_actor(request),
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/drafts")
async def create_draft(
    course_id: str,
    body: DraftRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "success",
            "workbench": await service.create_draft(
                course_id,
                actor=_actor(request),
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.patch("/drafts/{draft_id}")
async def patch_draft(
    course_id: str,
    draft_id: str,
    body: PatchDraftRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "success",
            "workbench": await service.patch_draft(
                course_id,
                actor=_actor(request),
                draft_id=draft_id,
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.delete("/drafts/{draft_id}")
async def discard_draft(
    course_id: str,
    draft_id: str,
    body: CommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "discarded",
            "workbench": await service.discard_draft(
                course_id,
                actor=_actor(request),
                draft_id=draft_id,
                idempotency_key=body.idempotency_key,
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/drafts/{draft_id}/ai-candidates")
async def create_ai_candidate(
    course_id: str,
    draft_id: str,
    body: AICandidateRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "success",
            "workbench": await service.create_ai_candidate(
                course_id,
                actor=_actor(request),
                draft_id=draft_id,
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/ai-candidates/{candidate_id}/accept")
async def accept_ai_candidate(
    course_id: str,
    candidate_id: str,
    body: CandidateCommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "accepted",
            "workbench": await service.accept_ai_candidate(
                course_id,
                actor=_actor(request),
                candidate_id=candidate_id,
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/ai-candidates/{candidate_id}/reject")
async def reject_ai_candidate(
    course_id: str,
    candidate_id: str,
    body: CommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "rejected",
            "workbench": await service.reject_ai_candidate(
                course_id,
                actor=_actor(request),
                candidate_id=candidate_id,
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/validate")
async def validate_draft(
    course_id: str,
    body: ChangeSetRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "success",
            "review": service.review_draft(course_id, actor=_actor(request), draft_id=body.draft_id),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/change-sets")
async def create_change_set(
    course_id: str,
    body: ChangeSetRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "success",
            "workbench": await service.create_change_set(
                course_id,
                actor=_actor(request),
                **body.model_dump(),
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/change-sets/{change_set_id}/apply")
async def apply_change_set(
    course_id: str,
    change_set_id: str,
    body: CommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "applied",
            **await service.apply_change_set(
                course_id,
                actor=_actor(request),
                change_set_id=change_set_id,
                idempotency_key=body.idempotency_key,
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/change-sets/{change_set_id}/reject")
async def reject_change_set(
    course_id: str,
    change_set_id: str,
    body: CommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "rejected",
            "workbench": await service.reject_change_set(
                course_id,
                actor=_actor(request),
                change_set_id=change_set_id,
                idempotency_key=body.idempotency_key,
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


@router.get("/revisions")
async def list_revisions(
    course_id: str,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {"status": "success", "revisions": service.view(course_id, actor=_actor(request))["revisions"]}
    except Exception as exc:
        raise _error(exc) from exc


@router.get("/revisions/{revision_id}/diff")
async def revision_diff(
    course_id: str,
    revision_id: str,
    against: str,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {"status": "success", **service.revision_diff(course_id, left=against, right=revision_id)}
    except Exception as exc:
        raise _error(exc) from exc


@router.post("/revisions/{revision_id}/restore")
async def restore_revision(
    course_id: str,
    revision_id: str,
    body: CommandRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "restored",
            **await service.restore_revision(
                course_id,
                actor=_actor(request),
                revision_id=revision_id,
                idempotency_key=body.idempotency_key,
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc


class RebuildRequest(CommandRequest):
    # 定向重建：不传就按整份下游状态里所有 rebuild_required 的对象来。
    only_types: list[str] = Field(default_factory=list, max_length=16)
    only_ids: list[str] = Field(default_factory=list, max_length=256)
    # 默认生成候选等教师确认，而不是直接覆盖正式产物。
    candidate_only: bool = True


@router.post("/downstream/rebuild")
async def rebuild_downstream(
    course_id: str,
    body: RebuildRequest,
    request: Request,
    service: TeachingPlanWorkbenchService = Depends(_service),
) -> dict[str, Any]:
    try:
        return {
            "status": "rebuilt",
            **await service.rebuild_downstream(
                course_id,
                actor=_actor(request),
                idempotency_key=body.idempotency_key,
                only_types=body.only_types or None,
                only_ids=body.only_ids or None,
                candidate_only=body.candidate_only,
            ),
        }
    except Exception as exc:
        raise _error(exc) from exc
